docs/adr/0003-graphql-layer-between-backend-frontend.md

```diff
@@ -0,0 +1,98 @@
+# GraphQL layer between backend and frontend
+
+Date: 2020-04-30
+
+## Status
+
+**Accepted**
+
+## Related tickets (not available for public)
+
+SIG-2568
+
+## Context
+
+Since our current backend only serves data through REST and there is no possibility to generate
+types and or interfaces, there is no also no automated way to validate data on the frontend
+which is coming from the backend.
```

Comment:
I think there is an [OpenAPI 3 specification](https://api.data.amsterdam.nl/signals/swagger/openapi.yaml) for the backend. Maybe [openapi-generator](https://github.com/OpenAPITools/openapi-generator) could be used to generate types and interfaces based on this specification?

Reply:
Great feedback, thanks!

I have used `openapi-generator` before and at least 3 or 4 tools like that, I even attempted and actually started writing my own and did some pull requests on some of comparable projects.

In the end they kept feeling like `bandages` and `incomplete hacks` and I IMHO they only work great for kind of simple projects or a project which is frozen. The moment I solved 'my challenges' like this with a `GraphQL` layer like in this proposal I never looked back.

Even if we would use one of those tools we still have to do lots of additional work on the frontend which I think we could solve cleaner and faster with this proposed `ADR`.

Reply:
I agree the tooling in the GraphQL ecosystem is superior to something like openapi-generator.